### DISCUSSION TITLE
[Idea] add the ability to associate features to each combination

### DISCUSSION THREAD

**Discussion by Maofree** [PrestaShop/PrestaShop#37479] [Ideas]
### The problem

Hi
a lack of prestashop is in the management of combinations, it would be useful to add the possibility of associating the features and descriptions to each combination.
I'll give you an example, a vase can have the diameter attribute, but by changing the diameter then usually the vase changes the height and width.

at the moment the features are common to all combinations but this is not true, it depends on the type of product.
usually the measurements could change but also the materials, not just the weight and price

the priority would be to be able to associate each combination with its own features, in some cases it could be useful to change the descriptions, but I find it a secondary thing.

I discovered that woocommerce has this basic function

on addons there are two modules that do this thing, but the problem is that they are not associated with the filter search module of ps_facetedsearch.
https://addons.prestashop.com/it/combinazioni-personalizzazione/39621-product-combination-features-combifeatures.html

however such an important thing should be added in the next versions of prestashop, given that you are making many changes in the code of prestashop 9 and later

this need is mainly related to satisfy the requests of external marketplaces like amazon, manomano and all the others.
Marketplaces manage combinations as single products, therefore they require specific characteristics for each combination.
pass values ​​​​such as
height: 100, 120, 140
width: 120, 140, 190
colors: yellow, red, green

they are not correct, because they expect single values ​​​​such as
height: 100
width: 120
color: red

### The solution

have two logics in the management of the features
- the current one as generic features for each combination and for products without combinations
- the second one to manage specific features for each combination. These would be added to the generic ones when the features of a particular combination are requested, both for the front part and for any external systems such as marketplace.
in the front part, when the combinations change, there should be the variation of the specific features for the selected combination.

in this way a backward compatibility with past systems and solutions is maintained, and a new solution would be added that will need new tables in the database.
so these features could be used only for complex products such as vases, lamps, shower boxes etc... and for all the others such as clothes and shoes they could avoid using them, using generic ones as is done now.

Obviously the facedsearch module will also have to be able to manage these new specific features

### Do you plan to work on this subject?

- [ ] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [ ] I'm willing to provide any wireframes or design assets required for this feature.
- [ ] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by jf-viguier** [comment]
It's a musts have. It should be in the core to have modules support it.

**Comment by Maofree** [comment]
a major problem of this lack, is with external portals and comparators. There are products with 100 combinations, with portals often a combination is considered as a single product, so prestashop always passes the same characteristics for each combination.
in lighting, variants with smaller dimensions could have fewer bulbs, but prestashop passes the same data for each combination, except only the price, references and weight

**Comment by bicipres** [comment]
Hi,

I agree, it really is a necessary feature that it would be perfect to integrate natively into Prestashop.

Example: We have lighting fixtures with a common base, but each fixture has its own characteristics that we've adapted. 

This means that each light has its own specific use, and it is therefore necessary to have characteristics linked to each combination.

In addition to the characteristics, the idea of having a description for each combination is for me to include in this subject and it is also necessary to have a unique title for each declination.

The module proposed above is indeed the only solution to date to enable us to have something functional, but in addition to the limitation with ps_faceted search, it is also limited on the functionality of bulk import of characteristics for combinations. The import of characteristics for each combination would have to be included in prestashop's bulk import. If this is not the case, then the API should be adapted accordingly.
Another point, while we're at it, is that if refactoring work is done on a unique title, description and characteristics for each declination, we could very well consider making a declination indexable with an index follow or no follow status for those who wish to extend the search engine visibility of their declinations.

At the very least, I think it's essential to be able to set up specific characteristics for each variation, if you want to be able to promote products on Google Shopping, which requires each product to have its own specific characteristics. In my opinion, this is a must-have in the context of PSmarketing, and therefore a choice for the future.

Regards,

**Comment by ivan-kiwik** [comment]
As a web agency, we often encounter this need. As already said, the modules on the market are not compatible with faceted filters. This is a must have!

**Comment by Hlavtox** [comment]
@Maofree @jf-viguier I definitely get your point guys, but I am afraid that it would complicate things a lot - from a technical standpoint.

Are there any other usecases for this? Because for example, for the vase, you can do the combinations like this:
Dimension:
- Diameter 20 cm, height 40 cm
- Diameter 10 cm, height 20 cm

Would it work for you?

----------

Or, just try to look on it from a different angle. Let me explain.

Usually, in most cases combinations are/should used if the only thing changing is the attribute, nothing else. For example:
- Shoe - Size - 39,40,41,42,43
- T-shirt - Size S,M,L + Color Red, Green

I think that if you have a product where also other features change depending on selected combination, it should really be separate products. It will index better in search engines, easier to manage.

**Comment by Maofree** [comment]
Hi @Hlavtox 
the problem occurs with products where if I change the width of a parameter, for example diameter, then the height and length or depth also change and this happens with products such as vases, but also with bathroom products or lamps, such as shower enclosures, the width changes and I can put only one variant, but then the characteristics of that product are no longer true
a 100 x 100cm shower enclosure has different characteristics from the 140 x 160cm one

this causes problems with external systems connected to prestashop, such as marketplaces that treat a combination as a single product and would like the characteristics referred to this combination, instead they always receive the same

values ​​such as
height: 100, 120, 140
width: 120, 140, 190
colors: yellow, red, green

they are not correct, because they expect single values ​​such as
height: 100
color: red

I know these things well by following various companies, but I imagine you do too know

get around this problem by creating individual products for each combination, it is not always possible, imagine shower boxes, which have many sizes and colors, there are products that have 100 combinations

I do not like wordpress and woocommerce at all, but it manages the characteristics for each combination, I saw it with a competitor of a client of mine who are friends and they compared the configurations

unfortunately beyond the difficulties in implementing this solution on prestashop, it is something requested and necessary for retailers in some sectors

with shoes or t-shirts this problem does not exist, because the size of the product is not important and the variant is managed by the attribute, so the exact value of the combination is passed to the marketplaces. Instead, for other products, changing the combination involves changing several parameters

you could use the current characteristics as generic characteristics (so you would not have to change anything compared to what is now in prestashop) and add characteristics for the combinations using new tables in the database, which are added to the general ones when the variants change, a bit like what happens with the references that change when the combination changes.
I think this solution is much simpler to manage and would not lead to changes in the current code except adding new solutions

using this solution is not good with marketplaces, because you pass two parameters in one and they want only one value
Diameter 20 cm, height 40 cm
however there would always be the problem of the third side that could vary for each combination as a consequence of the size of the product

Nowadays, sites are connected to other systems, and are also used as management systems since they manage various types of data, so they are mainly connected to marketplaces in various ways and since these manage combinations as single products, adding specific characteristics for each combination becomes essential.

I thought of a little while ago, in addition to the presence of generic characteristics common to all combinations, it is necessary to add characteristics for each combination, so redundancies would be avoided in the case of products such as shoes or t-shirts, which would not require specific data. In this way there would also be backward compatibility with all past systems and configurations

**Comment by Maofree** [comment]
in the end prestashop users are companies and each one has particular solutions, prestashop and we programmers in the end try to accommodate the needs of these companies, having customers happy with what we offer I think is the ideal situation to not have stress in our work.

I summarize what prestashop currently does not satisfy some companies, because it offers limited solutions.
- first of all are the characteristics in the combinations (as usually said there are common characteristics between all the combinations but there are others specific to each combination, even if in general they are few, but they are there)
- management of the shipping costs rates, because the couriers calculate the costs mainly based on the volume of the packages (there are also other parameters but the management also by volume should be considered)
- then the urls, because now I wanted to make some corrections in the logic of the urls and I see that there is no redirect logic if I wanted to make this type of change
from
{category:/}{id}{-:id_product_attribute}-{rewrite}{-:ean13}.html
to
{id}{-:id_product_attribute}-{rewrite}{-:ean13}.html

or from
{category:/}{id}{-:id_product_attribute}-{rewrite}{-:ean13}.html
to
{category:/}{rewrite}{-:ean13}-{id}{-:id_product_attribute}.html

use external modules that currently use override to work on the urls, I don't like them especially because I don't want to tie myself to modules made by others, where they are often not optimized or after a few years they end the development of the module. These are basic things that would give value to prestashop. Imagine a 10-15 year old site, maybe some time ago there were some seo logics and now others or the company wants to experiment with other solutions, there are some important aspects that could be improved

I will soon try to give my contribution also through PR not only with bug reports, I am organizing myself to do it

**Comment by jf-viguier** [comment]
We agree on committer monthly to do it on a stable feature flag

**Comment by Hlavtox** [comment]
@jf-viguier @Maofree Sorry, but my and @ShaiMagal opinion is that features are for things that are common for the product.

If some feature changes between combinations, it should be defined by it's attributes.

Can you provide some examples of cases where this is needed? I am working with Prestashop for over 10 years now and never had this need. 🤷‍♂️

**Reply by kpodemski** [reply]
There are plenty of such examples, and by default, merchants provide them in detail inside the description.

Attributes define what you choose (size, color, material), while features describe what you get as a result of that choice, so naturally there are some examples.

- A brake pad set is sold as a combination for different vehicle models. Each combination may have a different friction coefficient and operating temperature range.
- Candles in _small_ vs. _large_. Burn time is a feature: 20 hours for the small, 50 hours for the large.
- Batteries in a pack of 4 vs. a pack of 8. The total energy capacity feature differs per combination.

Of course, we could find a workaround on how to deal with it, but that doesn't mean the need isn't there ;)

**Comment by mitalterrecotte-create** [comment]
Hi @Hlavtox 
there are many cases where a company 

https://www.terrecottemital.it/shop/it/orci-in-terracotta/297-209-orcio-toscano-terracotta-impruneta.html
This feature is required.
We're talking about all cases where the features vary based on the product's available combinations.
The most obvious case is where there's a choice, for example, by width or diameter. Consequently, the various dimensions can change, but the decorations for each combination, or the materials, could also change.
Take this product as an example. It's a model where the height varies; obviously, the width and weight also change, which are managed as features.

https://www.mondialshop.it/it/piatti-doccia-in-marmo-resina/15775-50048-piatto-doccia-in-resina-bianco-effetto-cemento-3cm-star-plus-7438643341298.html
The same problem exists here because the combinations are based on size, but the height also varies accordingly, and there are other details that should be associated with the change in size. Most importantly, the weight changes, which here too is considered the weight of the product and not the shipping package.

https://www.kasa-store.com/it/tavoli-fissi/49-11949-riedizione-del-tavolo-tulip-di-saarinen-tondo-diam-cm-80-cm-90-cm-100-cm-107-o120-piano-laminato-o-in-marmo.html
Here, the need for product-related features and combination-related features is even more evident.
Here, in addition to the choice of material, there's the choice of diameter, but changing the diameter also changes the height, width, and weight of various table parts. In fact, they're forced to include this data in the descriptions, but that's not correct.

Another problem is sending precise data to external systems like portals, where each combination is managed like a product, but the characteristics of one combination are different from the others.

https://www.lampadaribartalini.it/incasso-e-segnapasso-moderne/3774-13613-faretto-incasso-led-moderno-pan-international-bart-est700-est701-acciaio-tondo-spot-8024598051512.html

The same problem also exists in the lighting sector,
for example, the choice of color temperature, which involves a change in luminous flux from 3000 to 4000°K.

This problem is especially present in furniture.

These are the companies that frequently complain about this problem and have objective difficulties with external systems in passing the specifications of individual combinations as well as displaying accurate data to customers on the website.

However, there are others as well.

**Comment by jf-viguier** [comment]
@Maofree I've tried a PR to add features to combinations, feel free to test it https://github.com/PrestaShop/PrestaShop/pull/41830

**Comment by Maofree** [comment]
Hi @jf-viguier 
thanks
in the next weekend I'll try it

**Comment by Maofree** [comment]
Hi @jf-viguier 
I saw that it has already been tested and approved for PrestaShop 9.3.0.
But the ps_facedsearch module doesn't seem to have been considered in the changes.
Will it also be able to manage filters for the features used in individual combinations?

**Comment by sophiacartersc263-droid** [comment]
This sounds like a useful improvement for products with many different combinations. Allowing each combination to have its own features would make product information more accurate and could also help when sending product data to external marketplaces. The idea of keeping the current system for simpler products also seems practical.

**Comment by sophiacartersc263-droid** [comment]
I agree that different product combinations can sometimes need their own details instead of sharing the same information. This would be especially useful for digital products where different versions may have different features or requirements. Having clear [product information](https://minecraftmod.org/) for each version can also make listings easier for users to understand.

**Comment by Maofree** [comment]
Hi @sophiacartersc263-droid 
thanks to @jf-viguier this feature will be available in prestashop 9.3.0
https://github.com/PrestaShop/PrestaShop/pull/41830

I don't remember if he also implemented the fixes in the ps_facedsearch module, maybe yes
